### PR TITLE
feat: MediaService takes media threshold for a given media format

### LINKED ISSUE #5395: [on hold because of carousel refactor][Android] Banner component not changing images on mobile devices
Banner component not changing its image on smaller screens ie different devices, especially **Android** devices? The image always stays the same, no matter what other images you have defined in **impex file**. 
This is replicable on https://spartacus.c39j2-walkersde1-d4-public.model-t.cc.commerce.ondemand.com/electronics-spa/en/USD/ as well.

**Original source:** 
https://stackoverflow.com/questions/58707944/banner-component-not-changing-images-on-mobile-devices

_No screenshots provided_

### LINKED ISSUE #5398: [Spike needed] Banner slots for extra small devices and breaking points not in sync with each other
**Original source:**
https://app.slack.com/client/TCT7GM9LL/CD16V16FR/thread/CD16V16FR-1571829130.137000

Spartacus seems to use its own breakpoints which are valued in the breakpoint.service as follows:
const DEFAULT_BREAKPOINTS = {
 [BREAKPOINT.xs]: 576,
 [BREAKPOINT.sm]: 768,
 [BREAKPOINT.md]: 992,
 [BREAKPOINT.lg]: 1200,
};
These breakpoints however are mixed between spartacus components.
For example:
Header slots have min-width: 992 and after that they are MD.
Media has max-width: 992 and after that they're LG.
To my understanding slots use the breakpoints as min-width and medias use it as max-width.
Is this a mistake in Spartacus and if not, is there a simple way to unify the breakpoint rules?
(also Bootstrap 4 uses different breakpoints for devices than Bootstrap 3, might there be some mixup of these or are they set one device width behind on purpose?
https://getbootstrap.com/docs/4.0/layout/overview/)

-----------

When we go belowdevice width:1055, spartacus uses slots for Extra small Devices (xs).
however spartacus also uses the same Device Width for Small (sm) images. So never are they in sync with each other.
[BREAKPOINT.md]: 1055= XS on header slots
[BREAKPOINT.md]: 1055= SM on Medias
I've attached a picture to visualize the issue.

B460A461

-----------

In the last example we had overwriten the standard breakpoints, but the issue still applies.
I can try to break this down more if needed and sorry for being so persistent about this.
In as simply as I can put it:
In hybrises media container we put 4 different sized pictures for all 4 device widths (xs, sm, md, lg).
With spartacus it is impossible to get the xs picture to show.

### LINKED ISSUE #6629: Media component responsive images are not working for Chrome
## Environment Details

- Spartacus:  1.4.0
- Browser: 79.0.3945.130 
- OS: Windows 10
- Device: Desktop

## Steps to Reproduce
Create smart edit component with responsive image (MediaComponent)

## Observed Results

Image doesn't change based on screen resolution. It works on other browsers

## Expected Results
Different image is shown for every resolution that has provided image.

## Offer to Fix
[Stack overflow thread](https://stackoverflow.com/questions/28155861/google-chrome-version-40-srcset-attribute-problems)

Since Chrome 40 srcset should not be used for showing different images for different resolutions.
Can use [picture](https://www.w3schools.com/tags/tag_picture.asp) HTML5 element instead.

### LINKED ISSUE #7229: MediaService improvements
We've seen various issues with the media service. Some of them are reported, some where mentioned during customer interaction. 

The main problem is that the media service is driven by breakpoints, which is actually wrong. The breakpoints are only taken the full screen into account, while media are often inside a smaller layout. With that setup, the `srcset` that is used in images aren't leveraged as much as we should. We should introduce a specific media configuration to map a so-called threshold to a media format (media formats are used in the backend to distinguish size-format for the same media). 
When the `srcset` is properly set on the img element, there's no longer the need for components to dictate a specific format, as the browser will handle the format based no the `srcset`. 

Another behaviour we've seen with new projects, is a mismatch of media formats in the backend and spartacus. This results currently in no media at all. This looks very bad and demotivates the dev. We should fallback to any of the available media in this case, by default the most optimised version or a random one if we can't figure this out. 

This ticket will fix several related bugs:  #6629, #5398, #5395

### PR DESCRIPTION
**Pull request by tobi-or-not-tobi**
The media service no longer depends on the breakpoints, and is now using configurable media size information. It will now resolve a media always, either by configuration, best match, or the first format, to avoid an empty image for banners or products.

It is no longer needed to pass in a specific format to the MediaComponent, but it won't break if you still do so. Spartacus will not pass in a format anymore. 

closes #7229, fixes #6629, fixes #5398, fixes #5395

### CODE REVIEW

File: projects/storefrontlib/src/shared/components/media/media.service.ts [hunk 1]
```diff
@@ -1,108 +1,126 @@
...
+      ? mediaContainer
+      : this.resolveMedia(mediaContainer as MediaContainer, format);
+
     return {
-      src: this.getMainImage(container, format),
-      srcset: this.getSrcSet(container),
-      alt: alt || this.getAlt(container, format),
+      src: this.resolveAbsoluteUrl(mainMedia?.url),
+      alt: alt || mainMedia?.altText,
+      srcset: this.resolveSrcSet(mediaContainer),
```
Comment (dunqan) [inline projects/storefrontlib/src/shared/components/media/media.service.ts:61]: It seems that specifying `srcset` without `sizes` works as implicit size 100%
We might consider adding `sizes` here (and in the media component)

Comment (tobi-or-not-tobi) [reply projects/storefrontlib/src/shared/components/media/media.service.ts:61]: I like to postpone the introduction of sizes. The main issue here is the relation with sizes and viewport, where as images are often used inside a smaller layout; the viewport size would not really contribute here. 

I've "fixed" this (well, ensure it's aligned with the current approach), by introducing a CSS rule (`object-fit: scale-down`) that guaranties that the product image will not be rendered larger than the actual image.

Comment (dunqan) [reply projects/storefrontlib/src/shared/components/media/media.service.ts:61]: 👍

File: projects/storefrontlib/src/shared/components/media/media.service.ts [hunk 2]
```diff
@@ -1,108 +1,133 @@
...
-        code: 'desktop',
-        threshold: this.breakpointService.getSize(BREAKPOINT.md),
-      },
-      {
-        code: 'widescreen',
-        threshold: this.breakpointService.getSize(BREAKPOINT.lg),
-      },
-    ];
+  ) {
+    this.sortFormats();
```
Comment (dunqan) [inline projects/storefrontlib/src/shared/components/media/media.service.ts]: As configuration might not be final at constructor time, we should consider either using `ConfigInitializerService.getStableConfig`, or making this logic to run at the first time it's actually needed

**Review by dunqan — APPROVED** [review]

**Pull request merged**